it/src/main/java/com/google/cloud/teleport/it/PerformanceBenchmarkingBase.java

```diff
@@ -0,0 +1,306 @@
+      rowContent.put("template_version", jobInfo.templateVersion());
+      rowContent.put("template_type", jobInfo.templateType());
+      rowContent.put("test_name", testName.getMethodName());
+      // Convert parameters map to list of table row since it's a repeated record
+      List<TableRow> parameterRows = new ArrayList<>();
+      for (String parameter : jobInfo.parameters().keySet()) {
+        TableRow row =
+            new TableRow().set("name", parameter).set("value", jobInfo.parameters().get(parameter));
+        parameterRows.add(row);
+      }
+      rowContent.put("parameters", parameterRows);
+      // Convert metrics map to list of table row since it's a repeated record
+      List<TableRow> metricRows = new ArrayList<>();
+      for (String metric : metrics.keySet()) {
+        TableRow row = new TableRow().set("name", metric).set("value", metrics.get(metric));
+        metricRows.add(row);
+      }
+      rowContent.put("metrics", metricRows);
+      bigQueryResourceManager.write(
+          TestProperties.exportTable(), RowToInsert.of("rowId", rowContent));
+    } catch (IllegalStateException e) {
+      LOG.warn("Unable to export results to datastore. ", e);
+    }
+  }
+
+  /**
+   * Checks if the input PCollection has the expected number of messages.
+   *
+   * @param jobId JobId of the job
+   * @param pcollectionName the input pcollection name
+   * @param expectedElements expected number of messages
+   * @return whether the input pcollection has the expected number of messages.
+   */
+  public boolean waitForNumMessages(String jobId, String pcollectionName, Long expectedElements) {
+    try {
+      // the element count metric always follows the pattern <pcollection name>-ElementCount
+      String metricName = pcollectionName.replace(".", "-") + "-ElementCount";
+      Double metric = dataflowClient.getMetric(PROJECT, REGION, jobId, metricName);
+      if ((metric != null) && (metric == (double) expectedElements)) {
+        return true;
+      }
+      LOG.info(
+          "Expected {} messages in input PCollection, but found {}.", expectedElements, metric);
+      return false;
+    } catch (Exception e) {
+      LOG.warn("Encountered error when trying to measure input elements. ", e);
+      return false;
+    }
+  }
+
+  /**
+   * Computes the metrics of the given job using dataflow and monitoring clients.
+   *
+   * @param jobInfo Job info of the job
+   * @param inputPcollection inpit pcollection of the dataflow job to query additional metrics
+   * @return metrics
+   * @throws IOException if there is an issue sending the request
+   * @throws ParseException if timestamp is inaccurate
+   * @throws InterruptedException thrown if thread is interruptedn
+   */
+  protected Map<String, Double> getMetrics(JobInfo jobInfo, String inputPcollection)
+      throws ParseException, InterruptedException, IOException {
+    // Metrics take upto 3 minutes to show up
+    Thread.sleep(Duration.ofMinutes(4).toMillis());
```

Comment:
Can we add some kind of checks to see if metrics are available? I'd suggest [Awaitility](http://www.awaitility.org/) - great library